reference/test/preconditioner/ilu.cpp

```diff
@@ -188,11 +183,13 @@ TEST_F(Ilu, SetsCorrectMatrices)
 
 TEST_F(Ilu, CanBeCopied)
 {
-    auto ilu = ilu_pre_factory->generate(l_factor, u_factor);
+    auto ilu = ilu_pre_factory->generate(l_u_composition);
     auto before_l_solver = ilu->get_l_solver();
     auto before_u_solver = ilu->get_u_solver();
     // The switch up of matrices is intentional, to make sure they are distinct!
-    auto copied = ilu_prec_type::build().on(exec)->generate(u_factor, l_factor);
+    auto u_l_composition = composition::create(u_factor, l_factor);
```

Comment:
This reminds me, what do we do when someone switches the L and U factors ? Or do we go with the philosophy of Garbage in, Garbage out ?

Reply:
Since I have no way of checking if it is an upper or a lower matrix, it kind of is `garbage in, garbage out`. Some solvers will handle it just fine, others might have a bigger problem with it.

Reply:
How are non-triangular matrices handled in the TRS? If you have functions to test for it, I use them here.
The problem I see is that the matrix format is totally unknown, so checking is quite complicated if you do not want to convert to a specific format, which takes time and potentially a lot of memory.

Reply:
Hm, I guess that makes sense. Additionally, the user probably has to take care of which solvers to use as well. Because, it does not really make sense to to use CG, FCG/ CGS here I guess. Maybe you can also add a note about that ?